tests/Unit/Utilities/CssDocumentTest.php

```diff
@@ -0,0 +1,403 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pelago\Emogrifier\Tests\Unit\Utilities;
+
+use Pelago\Emogrifier\Utilities\CssDocument;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @covers \Pelago\Emogrifier\Utilities\CssDocument
+ */
+final class CssDocumentTest extends TestCase
+{
+    /**
+     * @var string
+     */
+    private const VALID_AT_FONT_FACE_RULE = '
+        @font-face {
+          font-family: "Foo Sans";
+          src: url("/foo-sans.woff2") format("woff2");
+        }
+    ';
+
+    /**
+     * @test
+     *
+     * @param string $selector
+     *
+     * @dataProvider provideSelector
+     */
+    public function parsesSelector(string $selector): void
+    {
+        $css = $selector . '{ color: green; }';
+        $subject = new CssDocument($css);
+
+        $result = $subject->getStyleRulesData([]);
+
+        self::assertCount(1, $result);
+        $trimmedResultantSelector = \trim($result[0]['selectors']);
+        self::assertSame(\trim($selector), $trimmedResultantSelector);
+    }
+
+    /**
+     * @return array<string, array<int, string>>
+     */
+    public function provideSelector(): array
+    {
+        return [
+            'universal' => ['*'],
+            'type' => ['p'],
+            'class' => ['.classy'],
+            'ID' => ['#toc'],
+            'attribute (presence)' => ['[title]'],
+            'attribute (value match)' => ['[href=https://example.org]'],
+            'attribute (word match)' => ['[class~=logo]'],
+            'attribute (hyphenated prefix match)' => ['[lang|=de]'],
+            'attribute (prefix match)' => ['[href^=#]'],
+            'attribute (suffix match)' => ['[href$=.org]'],
+            'attribute (substring match)' => ['[href*=example]'],
+            'attribute (match with single quotes)' => ['[href=\'https://example.org\']'],
+            'attribute (match with double quotes)' => ['[href="https://example.org"]'],
+            'attribute (case insensitive match)' => ['[href*=example i]'],
+            'pseudo class' => [':hover'],
+            'pseudo element' => ['::after'],
+            'vendor pseudo class' => [':-webkit-autofill'],
+            'vendor pseudo element' => ['::-webkit-progress-bar'],
+            'combined' => ['p.classy'],
+            'descendant' => ['p .classy'],
+            'child' => ['p > .classy'],
+            'general sibling' => ['h1 ~ p'],
+            'adjacent sibling' => ['h1 + p'],
+            'list' => ['h1, h2'],
+            'with space after' => ['p '],
+            'with line feed after' => ["p\n"],
+            'with Windows line ending after' => ["p\r\n"],
+            'with TAB after' => ["p\t"],
+            'minified' => ['p>.classy'],
+            'with line feeds within' => ["p\n>\n.classy"],
```

Comment:
I propose we split the selectors and the formatting problems into two data providers.

Reply:
Done.